File: src/backend/typescript/az/src/services/azureService.ts
import { ResourceManagementClient } from '@azure/arm-resources';
import { StorageManagementClient } from '@azure/arm-storage';
import { DefaultAzureCredential, ClientSecretCredential, TokenCredential } from '@azure/identity';
import { 
  ResourceGroup, 
  CreateResourceGroupRequest, 
  UpdateResourceGroupRequest,
  StorageAccount,
  CreateStorageAccountRequest,
  UpdateStorageAccountRequest,
  StorageEndpoints,
  AzureErrorResponse 
} from '@/types/azure';
import config from '@/config/environment';
import { createLogger } from '@/config/logger';

export class AzureResourceError extends Error {
  constructor(
    message: string,
    public readonly code: string = 'AZURE_ERROR',
    public readonly statusCode: number = 500,
    public readonly azureError?: AzureErrorResponse
  ) {
    super(message);
    this.name = 'AzureResourceError';
  }
}

export class AzureService {
  private resourceClient: ResourceManagementClient;
  private storageClient: StorageManagementClient;
  private logger = createLogger();

  constructor(correlationId?: string) {
    this.logger = createLogger(correlationId);
    this.resourceClient = this.createResourceClient();
    this.storageClient = this.createStorageClient();
  }

  private createResourceClient(): ResourceManagementClient {
    try {
      let credential: TokenCredential;

      // Try service principal authentication first
      if (config.AZURE_CLIENT_ID && config.AZURE_CLIENT_SECRET && config.AZURE_TENANT_ID) {
        this.logger.info('Using service principal authentication');
        credential = new ClientSecretCredential(
          config.AZURE_TENANT_ID,
          config.AZURE_CLIENT_ID,
          config.AZURE_CLIENT_SECRET
        );
      } else {
        this.logger.info('Using default Azure credential');
        credential = new DefaultAzureCredential();
      }

      const client = new ResourceManagementClient(
        credential,
        config.AZURE_SUBSCRIPTION_ID
      );

      this.logger.info('Successfully created Azure resource client');
      return client;

    } catch (error) {
      this.logger.error('Failed to create Azure resource client', { error });
      throw new AzureResourceError(
        'Failed to initialize Azure connection',
        'INITIALIZATION_ERROR',
        503
      );
    }
  }

  private createStorageClient(): StorageManagementClient {
    try {
      let credential: TokenCredential;

      // Use the same credential as resource client
      if (config.AZURE_CLIENT_ID && config.AZURE_CLIENT_SECRET && config.AZURE_TENANT_ID) {
        this.logger.info('Using service principal authentication for storage client');
        credential = new ClientSecretCredential(
          config.AZURE_TENANT_ID,
          config.AZURE_CLIENT_ID,
          config.AZURE_CLIENT_SECRET
        );
      } else {
        this.logger.info('Using default Azure credential for storage client');
        credential = new DefaultAzureCredential();
      }

      const client = new StorageManagementClient(
        credential,
        config.AZURE_SUBSCRIPTION_ID
      );

      this.logger.info('Successfully created Azure storage client');
      return client;

    } catch (error) {
      this.logger.error('Failed to create Azure storage client', { error });
      throw new AzureResourceError(
        'Failed to initialize Azure storage connection',
        'INITIALIZATION_ERROR',
        503
      );
    }
  }

  async testConnection(): Promise<boolean> {
    try {
      this.logger.info('Testing Azure connection');
      const startTime = Date.now();
      
      // Try to list resource groups with a limit to test connectivity
      const iterator = this.resourceClient.resourceGroups.list({ top: 1 });
      await iterator.next();
      
      const responseTime = Date.now() - startTime;
      this.logger.info('Azure connectivity test successful', { responseTime });
      return true;
      
    } catch (error) {
      this.logger.error('Azure connectivity test failed', { error });
      return false;
    }
  }

  async listResourceGroups(): Promise<ResourceGroup[]> {
    try {
      this.logger.info('Listing resource groups');
      const resourceGroups: ResourceGroup[] = [];

      for await (const resourceGroup of this.resourceClient.resourceGroups.list()) {
        resourceGroups.push({
          id: resourceGroup.id ?? '',
          name: resourceGroup.name ?? '',
          location: resourceGroup.location ?? '',
          type: resourceGroup.type ?? 'Microsoft.Resources/resourceGroups',
          tags: resourceGroup.tags ?? {},
          managedBy: resourceGroup.managedBy,
          provisioningState: resourceGroup.properties?.provisioningState
        });
      }

      this.logger.info(`Found ${resourceGroups.length} resource groups`);
      return resourceGroups;

    } catch (error) {
      this.logger.error('Failed to list resource groups', { error });
      this.handleAzureError(error);
      throw error; // This won't be reached due to handleAzureError throwing
    }
  }

  async getResourceGroup(name: string): Promise<ResourceGroup> {
    try {
      this.logger.info(`Getting resource group: ${name}`);
      
      const resourceGroup = await this.resourceClient.resourceGroups.get(name);
      
      const result: ResourceGroup = {
        id: resourceGroup.id ?? '',
        name: resourceGroup.name ?? '',
        location: resourceGroup.location ?? '',
        type: resourceGroup.type ?? 'Microsoft.Resources/resourceGroups',
        tags: resourceGroup.tags ?? {},
        managedBy: resourceGroup.managedBy,
        provisioningState: resourceGroup.properties?.provisioningState
      };

      this.logger.info(`Successfully retrieved resource group: ${name}`);
      return result;

    } catch (error) {
      this.logger.error(`Failed to get resource group ${name}`, { error });
      this.handleAzureError(error, name);
      throw error;
    }
  }

  async createResourceGroup(request: CreateResourceGroupRequest): Promise<ResourceGroup> {
    try {
      this.logger.info(`Creating resource group: ${request.name}`);
      
      const parameters = {
        location: request.location,
        tags: request.tags ?? {}
      };

      const result = await this.resourceClient.resourceGroups.createOrUpdate(
        request.name,
        parameters
      );

      const resourceGroup: ResourceGroup = {
        id: result.id ?? '',
        name: result.name ?? '',
        location: result.location ?? '',
        type: result.type ?? 'Microsoft.Resources/resourceGroups',
        tags: result.tags ?? {},
        managedBy: result.managedBy,
        provisioningState: result.properties?.provisioningState
      };

      this.logger.info(`Successfully created resource group: ${request.name}`);
      return resourceGroup;

    } catch (error) {
      this.logger.error(`Failed to create resource group ${request.name}`, { error });
      this.handleAzureError(error, request.name);
      throw error;
    }
  }

  async updateResourceGroup(name: string, request: UpdateResourceGroupRequest): Promise<ResourceGroup> {
    try {
      this.logger.info(`Updating resource group: ${name}`);
      
      // First check if resource group exists
      await this.getResourceGroup(name);
      
      const parameters = {
        tags: request.tags ?? {}
      };

      const result = await this.resourceClient.resourceGroups.update(name, parameters);

      const resourceGroup: ResourceGroup = {
        id: result.id ?? '',
        name: result.name ?? '',
        location: result.location ?? '',
        type: result.type ?? 'Microsoft.Resources/resourceGroups',
        tags: result.tags ?? {},
        managedBy: result.managedBy,
        provisioningState: result.properties?.provisioningState
      };

      this.logger.info(`Successfully updated resource group: ${name}`);
      return resourceGroup;

    } catch (error) {
      this.logger.error(`Failed to update resource group ${name}`, { error });
      this.handleAzureError(error, name);
      throw error;
    }
  }

  async deleteResourceGroup(name: string): Promise<void> {
    try {
      this.logger.info(`Deleting resource group: ${name}`);
      
      // First check if resource group exists
      await this.getResourceGroup(name);
      
      // Start the deletion (this is an async operation in Azure)
      const deletePoller = await this.resourceClient.resourceGroups.beginDelete(name);
      
      this.logger.info(`Started deletion of resource group: ${name}`);
      // Note: We don't wait for completion as it can take a long time
      
    } catch (error) {
      this.logger.error(`Failed to delete resource group ${name}`, { error });
      this.handleAzureError(error, name);
      throw error;
    }
  }

  // Storage Account Methods

  async listStorageAccounts(): Promise<StorageAccount[]> {
    try {
      this.logger.info('Listing storage accounts');
      const storageAccounts: StorageAccount[] = [];

      for await (const account of this.storageClient.storageAccounts.list()) {
        const resourceGroup = this.extractResourceGroupFromId(account.id ?? '');
        const storageAccount: StorageAccount = {
          id: account.id ?? '',
          name: account.name ?? '',
          location: account.location ?? '',
          resourceGroup,
          kind: account.kind ?? 'StorageV2',
          skuName: account.sku?.name ?? 'Standard_LRS',
          skuTier: account.sku?.tier,
          accessTier: account.accessTier,
          allowBlobPublicAccess: account.allowBlobPublicAccess,
          allowSharedKeyAccess: account.allowSharedKeyAccess,
          tags: account.tags ?? {},
          creationTime: account.creationTime?.toISOString(),
          primaryEndpoints: account.primaryEndpoints ? {
            blob: account.primaryEndpoints.blob,
            queue: account.primaryEndpoints.queue,
            table: account.primaryEndpoints.table,
            file: account.primaryEndpoints.file
          } : undefined
        };
        storageAccounts.push(storageAccount);
      }

      this.logger.info(`Found ${storageAccounts.length} storage accounts`);
      return storageAccounts;

    } catch (error) {
      this.logger.error('Failed to list storage accounts', { error });
      this.handleAzureError(error);
      throw error;
    }
  }

  async getStorageAccount(resourceGroup: string, name: string): Promise<StorageAccount> {
    try {
      this.logger.info(`Getting storage account: ${name} in resource group: ${resourceGroup}`);
      
      const account = await this.storageClient.storageAccounts.getProperties(resourceGroup, name);
      
      const storageAccount: StorageAccount = {
        id: account.id ?? '',
        name: account.name ?? '',
        location: account.location ?? '',
        resourceGroup,
        kind: account.kind ?? 'StorageV2',
        skuName: account.sku?.name ?? 'Standard_LRS',
        skuTier: account.sku?.tier,
        accessTier: account.accessTier,
        allowBlobPublicAccess: account.allowBlobPublicAccess,
        allowSharedKeyAccess: account.allowSharedKeyAccess,
        tags: account.tags ?? {},
        creationTime: account.creationTime?.toISOString(),
        primaryEndpoints: account.primaryEndpoints ? {
          blob: account.primaryEndpoints.blob,
          queue: account.primaryEndpoints.queue,
          table: account.primaryEndpoints.table,
          file: account.primaryEndpoints.file
        } : undefined
      };

      this.logger.info(`Successfully retrieved storage account: ${name}`);
      return storageAccount;

    } catch (error) {
      this.logger.error(`Failed to get storage account ${name}`, { error });
      this.handleAzureError(error, name);
      throw error;
    }
  }

  async createStorageAccount(request: CreateStorageAccountRequest): Promise<StorageAccount> {
    try {
      this.logger.info(`Creating storage account: ${request.name}`);
      
      const parameters = {
        location: request.location,
        kind: request.kind || 'StorageV2',
        sku: {
          name: request.skuName || 'Standard_LRS'
        },
        properties: {
          accessTier: request.accessTier,
          allowBlobPublicAccess: request.allowBlobPublicAccess ?? false,
          allowSharedKeyAccess: request.allowSharedKeyAccess ?? true
        },
        tags: request.tags ?? {}
      };

      // Start creation (this is an async operation)
      const createPoller = await this.storageClient.storageAccounts.beginCreate(
        request.resourceGroup,
        request.name,
        parameters
      );

      // Wait for completion
      const result = await createPoller.pollUntilDone();

      const storageAccount: StorageAccount = {
        id: result.id ?? '',
        name: result.name ?? '',
        location: result.location ?? '',
        resourceGroup: request.resourceGroup,
        kind: result.kind ?? 'StorageV2',
        skuName: result.sku?.name ?? 'Standard_LRS',
        skuTier: result.sku?.tier,
        accessTier: result.accessTier,
        allowBlobPublicAccess: result.allowBlobPublicAccess,
        allowSharedKeyAccess: result.allowSharedKeyAccess,
        tags: result.tags ?? {},
        creationTime: result.creationTime?.toISOString(),
        primaryEndpoints: result.primaryEndpoints ? {
          blob: result.primaryEndpoints.blob,
          queue: result.primaryEndpoints.queue,
          table: result.primaryEndpoints.table,
          file: result.primaryEndpoints.file
        } : undefined
      };

      this.logger.info(`Successfully created storage account: ${request.name}`);
      return storageAccount;

    } catch (error) {
      this.logger.error(`Failed to create storage account ${request.name}`, { error });
      this.handleAzureError(error, request.name);
      throw error;
    }
  }

  async updateStorageAccount(resourceGroup: string, name: string, request: UpdateStorageAccountRequest): Promise<StorageAccount> {
    try {
      this.logger.info(`Updating storage account: ${name}`);
      
      // First check if storage account exists
      await this.getStorageAccount(resourceGroup, name);
      
      const parameters = {
        properties: {
          accessTier: request.accessTier,
          allowBlobPublicAccess: request.allowBlobPublicAccess,
          allowSharedKeyAccess: request.allowSharedKeyAccess
        },
        tags: request.tags
      };

      const result = await this.storageClient.storageAccounts.update(resourceGroup, name, parameters);

      const storageAccount: StorageAccount = {
        id: result.id ?? '',
        name: result.name ?? '',
        location: result.location ?? '',
        resourceGroup,
        kind: result.kind ?? 'StorageV2',
        skuName: result.sku?.name ?? 'Standard_LRS',
        skuTier: result.sku?.tier,
        accessTier: result.accessTier,
        allowBlobPublicAccess: result.allowBlobPublicAccess,
        allowSharedKeyAccess: result.allowSharedKeyAccess,
        tags: result.tags ?? {},
        creationTime: result.creationTime?.toISOString(),
        primaryEndpoints: result.primaryEndpoints ? {
          blob: result.primaryEndpoints.blob,
          queue: result.primaryEndpoints.queue,
          table: result.primaryEndpoints.table,
          file: result.primaryEndpoints.file
        } : undefined
      };

      this.logger.info(`Successfully updated storage account: ${name}`);
      return storageAccount;

    } catch (error) {
      this.logger.error(`Failed to update storage account ${name}`, { error });
      this.handleAzureError(error, name);
      throw error;
    }
  }

  async deleteStorageAccount(resourceGroup: string, name: string): Promise<void> {
    try {
      this.logger.info(`Deleting storage account: ${name}`);
      
      // First check if storage account exists
      await this.getStorageAccount(resourceGroup, name);
      
      // Delete the storage account
      await this.storageClient.storageAccounts.delete(resourceGroup, name);
      
      this.logger.info(`Successfully deleted storage account: ${name}`);
      
    } catch (error) {
      this.logger.error(`Failed to delete storage account ${name}`, { error });
      this.handleAzureError(error, name);
      throw error;
    }
  }

  private extractResourceGroupFromId(resourceId: string): string {
    if (!resourceId) return '';
    
    const parts = resourceId.split('/');
    const rgIndex = parts.findIndex(part => part === 'resourceGroups');
    
    if (rgIndex !== -1 && rgIndex + 1 < parts.length) {
      return parts[rgIndex + 1];
    }
    
    return '';
  }

  private handleAzureError(error: any, resourceName?: string): never {
    // Handle different types of Azure errors
    if (error?.code === 'ResourceGroupNotFound' || error?.statusCode === 404) {
      throw new AzureResourceError(
        `Resource group${resourceName ? ` '${resourceName}'` : ''} not found`,
        'RESOURCE_NOT_FOUND',
        404
      );
    }
    
    if (error?.code === 'ResourceGroupBeingDeleted') {
      throw new AzureResourceError(
        `Resource group '${resourceName}' is being deleted`,
        'RESOURCE_BEING_DELETED',
        409
      );
    }

    if (error?.statusCode === 403 || error?.code?.includes('Authorization')) {
      throw new AzureResourceError(
        'Insufficient permissions to perform this operation',
        'AUTHORIZATION_ERROR',
        403
      );
    }

    if (error?.statusCode === 409 || error?.code?.includes('Conflict')) {
      throw new AzureResourceError(
        `Resource group '${resourceName}' already exists or has a conflict`,
        'CONFLICT_ERROR',
        409
      );
    }

    // Handle rate limiting
    if (error?.statusCode === 429) {
      throw new AzureResourceError(
        'Too many requests to Azure API, please try again later',
        'RATE_LIMIT_ERROR',
        429
      );
    }

    // Generic Azure error
    const message = error?.message || error?.body?.message || 'An Azure operation failed';
    const code = error?.code || 'AZURE_ERROR';
    const statusCode = error?.statusCode || 500;

    throw new AzureResourceError(message, code, statusCode);
  }
}

File: src/backend/python/az/api/services/azure_service.py
from typing import List, Optional

from azure.core.exceptions import ResourceNotFoundError, HttpResponseError
from azure.identity import DefaultAzureCredential, ClientSecretCredential
from azure.mgmt.resource import ResourceManagementClient
from azure.mgmt.storage import StorageManagementClient

from api.core.config import Settings
from api.core.logging import get_logger
from api.models.resource_group import (
    ResourceGroup, ResourceGroupCreate, ResourceGroupUpdate,
    StorageAccount, StorageAccountCreate, StorageAccountUpdate, StorageEndpoints
)


logger = get_logger(__name__)


class AzureService:
    """Azure service for managing Azure resources."""
    
    def __init__(self, settings: Settings):
        self.settings = settings
        self.resource_client = self._create_resource_client()
        self.storage_client = self._create_storage_client()
    
    def _create_resource_client(self) -> ResourceManagementClient:
        """Create Azure resource management client with proper authentication."""
        try:
            # Try service principal authentication first
            if all([
                self.settings.azure_client_id,
                self.settings.azure_client_secret,
                self.settings.azure_tenant_id
            ]):
                logger.info("Using service principal authentication")
                credential = ClientSecretCredential(
                    tenant_id=self.settings.azure_tenant_id,
                    client_id=self.settings.azure_client_id,
                    client_secret=self.settings.azure_client_secret
                )
            else:
                # Fall back to default credential (managed identity, Azure CLI, etc.)
                logger.info("Using default Azure credential")
                credential = DefaultAzureCredential()
            
            client = ResourceManagementClient(
                credential=credential,
                subscription_id=self.settings.azure_subscription_id
            )
            
            logger.info("Successfully created Azure resource client")
            return client
            
        except Exception as e:
            logger.error(f"Failed to create Azure resource client: {e}")
            raise
    
    def _create_storage_client(self) -> StorageManagementClient:
        """Create Azure storage management client with proper authentication."""
        try:
            # Use the same credential as resource client
            if all([
                self.settings.azure_client_id,
                self.settings.azure_client_secret,
                self.settings.azure_tenant_id
            ]):
                logger.info("Using service principal authentication for storage client")
                credential = ClientSecretCredential(
                    tenant_id=self.settings.azure_tenant_id,
                    client_id=self.settings.azure_client_id,
                    client_secret=self.settings.azure_client_secret
                )
            else:
                logger.info("Using default Azure credential for storage client")
                credential = DefaultAzureCredential()
            
            client = StorageManagementClient(
                credential=credential,
                subscription_id=self.settings.azure_subscription_id
            )
            
            logger.info("Successfully created Azure storage client")
            return client
            
        except Exception as e:
            logger.error(f"Failed to create Azure storage client: {e}")
            raise
    
    async def test_connection(self) -> bool:
        """Test Azure connection."""
        try:
            # Try to list resource groups (limited to 1) to test connectivity
            list(self.resource_client.resource_groups.list(top=1))
            logger.info("Azure connectivity test successful")
            return True
        except Exception as e:
            logger.error(f"Azure connectivity test failed: {e}")
            return False
    
    async def list_resource_groups(self) -> List[ResourceGroup]:
        """List all resource groups in the subscription."""
        try:
            logger.info("Listing resource groups")
            resource_groups = []
            
            for rg in self.resource_client.resource_groups.list():
                resource_groups.append(ResourceGroup(
                    id=rg.id,
                    name=rg.name,
                    location=rg.location,
                    tags=rg.tags or {},
                    type=rg.type,
                    managed_by=rg.managed_by
                ))
            
            logger.info(f"Found {len(resource_groups)} resource groups")
            return resource_groups
            
        except Exception as e:
            logger.error(f"Failed to list resource groups: {e}")
            raise
    
    async def get_resource_group(self, name: str) -> ResourceGroup:
        """Get a specific resource group by name."""
        try:
            logger.info(f"Getting resource group: {name}")
            
            rg = self.resource_client.resource_groups.get(name)
            
            return ResourceGroup(
                id=rg.id,
                name=rg.name,
                location=rg.location,
                tags=rg.tags or {},
                type=rg.type,
                managed_by=rg.managed_by
            )
            
        except ResourceNotFoundError:
            logger.warning(f"Resource group not found: {name}")
            raise
        except Exception as e:
            logger.error(f"Failed to get resource group {name}: {e}")
            raise
    
    async def create_resource_group(self, resource_group: ResourceGroupCreate) -> ResourceGroup:
        """Create a new resource group."""
        try:
            logger.info(f"Creating resource group: {resource_group.name}")
            
            parameters = {
                "location": resource_group.location,
                "tags": resource_group.tags or {}
            }
            
            result = self.resource_client.resource_groups.create_or_update(
                resource_group.name,
                parameters
            )
            
            logger.info(f"Successfully created resource group: {resource_group.name}")
            
            return ResourceGroup(
                id=result.id,
                name=result.name,
                location=result.location,
                tags=result.tags or {},
                type=result.type,
                managed_by=result.managed_by
            )
            
        except HttpResponseError as e:
            logger.error(f"Azure API error creating resource group {resource_group.name}: {e}")
            raise
        except Exception as e:
            logger.error(f"Failed to create resource group {resource_group.name}: {e}")
            raise
    
    async def update_resource_group(self, name: str, update: ResourceGroupUpdate) -> ResourceGroup:
        """Update an existing resource group."""
        try:
            logger.info(f"Updating resource group: {name}")
            
            # First check if resource group exists
            await self.get_resource_group(name)
            
            # Update only tags (location cannot be changed)
            parameters = {
                "tags": update.tags or {}
            }
            
            result = self.resource_client.resource_groups.update(name, parameters)
            
            logger.info(f"Successfully updated resource group: {name}")
            
            return ResourceGroup(
                id=result.id,
                name=result.name,
                location=result.location,
                tags=result.tags or {},
                type=result.type,
                managed_by=result.managed_by
            )
            
        except ResourceNotFoundError:
            logger.warning(f"Cannot update non-existent resource group: {name}")
            raise
        except Exception as e:
            logger.error(f"Failed to update resource group {name}: {e}")
            raise
    
    async def delete_resource_group(self, name: str) -> None:
        """Delete a resource group."""
        try:
            logger.info(f"Deleting resource group: {name}")
            
            # Check if resource group exists first
            await self.get_resource_group(name)
            
            # Start the deletion (this is an async operation in Azure)
            delete_operation = self.resource_client.resource_groups.begin_delete(name)
            
            logger.info(f"Started deletion of resource group: {name}")
            # Note: We don't wait for completion as it can take a long time
            
        except ResourceNotFoundError:
            logger.warning(f"Cannot delete non-existent resource group: {name}")
            raise
        except Exception as e:
            logger.error(f"Failed to delete resource group {name}: {e}")
            raise
    
    # Storage Account Methods
    
    async def list_storage_accounts(self) -> List[StorageAccount]:
        """List all storage accounts in the subscription."""
        try:
            logger.info("Listing storage accounts")
            storage_accounts = []
            
            for sa in self.storage_client.storage_accounts.list():
                # Extract resource group from ID
                resource_group = self._extract_resource_group_from_id(sa.id)
                
                endpoints = None
                if sa.primary_endpoints:
                    endpoints = StorageEndpoints(
                        blob=sa.primary_endpoints.blob,
                        queue=sa.primary_endpoints.queue,
                        table=sa.primary_endpoints.table,
                        file=sa.primary_endpoints.file
                    )
                
                storage_accounts.append(StorageAccount(
                    id=sa.id,
                    name=sa.name,
                    location=sa.location,
                    resource_group=resource_group,
                    kind=sa.kind,
                    sku_name=sa.sku.name if sa.sku else None,
                    sku_tier=sa.sku.tier if sa.sku else None,
                    access_tier=sa.access_tier,
                    allow_blob_public=sa.allow_blob_public_access,
                    allow_shared_key=sa.allow_shared_key_access,
                    tags=sa.tags or {},
                    creation_time=sa.creation_time,
                    primary_endpoints=endpoints
                ))
            
            logger.info(f"Found {len(storage_accounts)} storage accounts")
            return storage_accounts
            
        except Exception as e:
            logger.error(f"Failed to list storage accounts: {e}")
            raise
    
    async def get_storage_account(self, resource_group: str, name: str) -> StorageAccount:
        """Get a specific storage account by name."""
        try:
            logger.info(f"Getting storage account: {name} in resource group: {resource_group}")
            
            sa = self.storage_client.storage_accounts.get_properties(resource_group, name)
            
            endpoints = None
            if sa.primary_endpoints:
                endpoints = StorageEndpoints(
                    blob=sa.primary_endpoints.blob,
                    queue=sa.primary_endpoints.queue,
                    table=sa.primary_endpoints.table,
                    file=sa.primary_endpoints.file
                )
            
            return StorageAccount(
                id=sa.id,
                name=sa.name,
                location=sa.location,
                resource_group=resource_group,
                kind=sa.kind,
                sku_name=sa.sku.name if sa.sku else None,
                sku_tier=sa.sku.tier if sa.sku else None,
                access_tier=sa.access_tier,
                allow_blob_public=sa.allow_blob_public_access,
                allow_shared_key=sa.allow_shared_key_access,
                tags=sa.tags or {},
                creation_time=sa.creation_time,
                primary_endpoints=endpoints
            )
            
        except ResourceNotFoundError:
            logger.warning(f"Storage account not found: {name}")
            raise
        except Exception as e:
            logger.error(f"Failed to get storage account {name}: {e}")
            raise
    
    async def create_storage_account(self, storage_account: StorageAccountCreate) -> StorageAccount:
        """Create a new storage account."""
        try:
            logger.info(f"Creating storage account: {storage_account.name}")
            
            from azure.mgmt.storage.models import (
                StorageAccountCreateParameters, Sku, Kind, AccessTier
            )
            
            # Set defaults and validate
            kind = Kind.STORAGE_V2
            if storage_account.kind:
                kind = getattr(Kind, storage_account.kind.upper().replace('V', '_V'), Kind.STORAGE_V2)
            
            sku = Sku(name=storage_account.sku_name or "Standard_LRS")
            
            parameters = StorageAccountCreateParameters(
                location=storage_account.location,
                kind=kind,
                sku=sku,
                tags=storage_account.tags or {}
            )
            
            # Set optional parameters
            if storage_account.access_tier:
                parameters.access_tier = getattr(AccessTier, storage_account.access_tier.upper(), None)
            
            if storage_account.allow_blob_public is not None:
                parameters.allow_blob_public_access = storage_account.allow_blob_public
                
            if storage_account.allow_shared_key is not None:
                parameters.allow_shared_key_access = storage_account.allow_shared_key
            
            # Start creation (this is an async operation)
            create_operation = self.storage_client.storage_accounts.begin_create(
                storage_account.resource_group,
                storage_account.name,
                parameters
            )
            
            # Wait for completion
            result = create_operation.result()
            
            logger.info(f"Successfully created storage account: {storage_account.name}")
            
            endpoints = None
            if result.primary_endpoints:
                endpoints = StorageEndpoints(
                    blob=result.primary_endpoints.blob,
                    queue=result.primary_endpoints.queue,
                    table=result.primary_endpoints.table,
                    file=result.primary_endpoints.file
                )
            
            return StorageAccount(
                id=result.id,
                name=result.name,
                location=result.location,
                resource_group=storage_account.resource_group,
                kind=result.kind,
                sku_name=result.sku.name if result.sku else None,
                sku_tier=result.sku.tier if result.sku else None,
                access_tier=result.access_tier,
                allow_blob_public=result.allow_blob_public_access,
                allow_shared_key=result.allow_shared_key_access,
                tags=result.tags or {},
                creation_time=result.creation_time,
                primary_endpoints=endpoints
            )
            
        except HttpResponseError as e:
            logger.error(f"Azure API error creating storage account {storage_account.name}: {e}")
            raise
        except Exception as e:
            logger.error(f"Failed to create storage account {storage_account.name}: {e}")
            raise
    
    async def update_storage_account(self, resource_group: str, name: str, update: StorageAccountUpdate) -> StorageAccount:
        """Update an existing storage account."""
        try:
            logger.info(f"Updating storage account: {name}")
            
            from azure.mgmt.storage.models import StorageAccountUpdateParameters, AccessTier
            
            # First check if storage account exists
            await self.get_storage_account(resource_group, name)
            
            parameters = StorageAccountUpdateParameters()
            
            if update.tags is not None:
                parameters.tags = update.tags
                
            if update.access_tier:
                parameters.access_tier = getattr(AccessTier, update.access_tier.upper(), None)
                
            if update.allow_blob_public is not None:
                parameters.allow_blob_public_access = update.allow_blob_public
                
            if update.allow_shared_key is not None:
                parameters.allow_shared_key_access = update.allow_shared_key
            
            result = self.storage_client.storage_accounts.update(resource_group, name, parameters)
            
            logger.info(f"Successfully updated storage account: {name}")
            
            endpoints = None
            if result.primary_endpoints:
                endpoints = StorageEndpoints(
                    blob=result.primary_endpoints.blob,
                    queue=result.primary_endpoints.queue,
                    table=result.primary_endpoints.table,
                    file=result.primary_endpoints.file
                )
            
            return StorageAccount(
                id=result.id,
                name=result.name,
                location=result.location,
                resource_group=resource_group,
                kind=result.kind,
                sku_name=result.sku.name if result.sku else None,
                sku_tier=result.sku.tier if result.sku else None,
                access_tier=result.access_tier,
                allow_blob_public=result.allow_blob_public_access,
                allow_shared_key=result.allow_shared_key_access,
                tags=result.tags or {},
                creation_time=result.creation_time,
                primary_endpoints=endpoints
            )
            
        except ResourceNotFoundError:
            logger.warning(f"Cannot update non-existent storage account: {name}")
            raise
        except Exception as e:
            logger.error(f"Failed to update storage account {name}: {e}")
            raise
    
    async def delete_storage_account(self, resource_group: str, name: str) -> None:
        """Delete a storage account."""
        try:
            logger.info(f"Deleting storage account: {name}")
            
            # Check if storage account exists first
            await self.get_storage_account(resource_group, name)
            
            # Delete the storage account
            self.storage_client.storage_accounts.delete(resource_group, name)
            
            logger.info(f"Successfully deleted storage account: {name}")
            
        except ResourceNotFoundError:
            logger.warning(f"Cannot delete non-existent storage account: {name}")
            raise
        except Exception as e:
            logger.error(f"Failed to delete storage account {name}: {e}")
            raise
    
    def _extract_resource_group_from_id(self, resource_id: str) -> str:
        """Extract resource group name from Azure resource ID."""
        if not resource_id:
            return ""
        
        parts = resource_id.split('/')
        try:
            # Azure resource ID format: /subscriptions/{sub}/resourceGroups/{rg}/...
            rg_index = parts.index('resourceGroups')
            if rg_index + 1 < len(parts):
                return parts[rg_index + 1]
        except (ValueError, IndexError):
            pass
        
        return ""